GNUmakefile

```diff
@@ -0,0 +1,76 @@
+# User-customizable variables:
+CPPFLAGS ?=
```

Comment:
This looks a bit over-engineered. Do we really need that many user knobs?
I would leave just CFLAGS and LDFLAGS for user, it is sufficient and enough.
TARGET_ARCH can be specified in CFLAGS. We can't have a separate var for just any compiler feature. E.g. if user wants to override warnings flags, we introduce WARNFLAGS?
LDLIBS and LOADLIBES, I don't understand difference between these two and between LDFLAGS. Libraries can always be added to LDFLAGS, no?
Supporting both CPPFLAGS and CXXFLAGS looks excessive.
Why do we need to parametrize OUTPUT_OPTION? Why not just say -o where needed?

Reply:
> This looks a bit over-engineered. Do we really need that many user knobs?

Good question! I do not need all these knobs; most of them are unnecessary for my needs.

I'm just supporting the variables which GNU `make` supports out-of-the-box if you don't have a makefile at all:

```
$ mkdir tmp
$ cd tmp
$ echo 'int main() {}' >hello.cpp
$ CXX=/usr/bin/clang++ CXXFLAGS=-std=c++11 LDFLAGS=-L/ LDLIBS=-lm TARGET_ARCH=-m32 CPPFLAGS=-I/ make -B hello
/usr/bin/clang++ -std=c++11 -I/ -L/ -m32 hello.cpp  -lm -o hello
$ ls -la
total 24
drwxr-xr-x   4 mg  staff   136 Nov 26 21:56 .
drwxr-xr-x  14 mg  staff   476 Nov 26 21:53 ..
-rwxr-xr-x   1 mg  staff  4228 Nov 26 21:56 hello
-rw-r--r--   1 mg  staff    14 Nov 26 21:56 hello.cpp
$ make --version
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
```

(`DEPFLAGS` is not used by built-in rules. I added `DEPFLAGS` specifically for this project.)

I would use GNU `make`'s built-in rules without recreating them, but I couldn't see how to do that with a build directory.

---

>  LDLIBS and LOADLIBES, I don't understand difference between these two and between LDFLAGS. Libraries can always be added to LDFLAGS, no?

`LDLIBS` is placed after the .o files, whereas `LDFLAGS` is placed before the .o files. The distinction is explained in GNU `make`'s documentation for the [linking built-in rule](https://www.gnu.org/software/make/manual/html_node/Catalogue-of-Rules.html#index-ld), and is hinted at in GNU `make`'s documentation for the [`LDFLAGS` and `LDLIBS` flags](https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html#index-LDFLAGS).

---

Let me know if you still want me to remove most of these variables.

Reply:
Yes.
Let's start simple and do only what we actually need. Which is no variables at all besides probably CFLAGS/LDFLAGS as they are the most common and well known.
FWIW the doc you linked says LOADLIBES is deprecated.
We can make this arbitrary complex later.